text/libobject-refactor.md

```diff
@@ -0,0 +1,228 @@
+  sequentially "load" the objects, basically performing the
+  side-effects declared in their interface
+
+- the objects are then stored in a internal map, as their interface
+  may need to be called when importing for example
+
+- users of `Declaremod` have to be careful about the state
+
+# Logical and meta-logical objects
+
+A common discussion topic is about the role that objects registered
+with the kernel play. Indeed, some objects present in a `.vo` can be
+understood as "logical" objects, as they will be registered with the
+kernel.
+
+A large set of objects are meta-logical: they are not registered with
+the kernel, but still essential to Coq operation. Such objects are
+notations, instance declarations, hints, etc...
+
+Indeed it is a common source of discussion where the data for logical
+objects should be stored. Some data will be stored invariant in the
+kernel, but some other data, like documentation or location, has to
+live outside it.
+
+# Part I: common meta-data
+
+For this part, we propose to decorate the `Libobject.t` type with a
+`Data.t` structure. This approach is modelled after the approach used
+in `Declare` for constant information, which has proven successful in
+reducing code duplication, uniformizing the API, and making it more
+extensible (see for example per-definition typing flags).
+
+In this preliminary proposal, `Data.t`, will contain:
+
+- `loc`: Location of the object, if it is coming from a source file
+- `doc`: String containing the documentation of the object, to be extended
+- `name`: Name of the object, as an `Id.t`, if not anonymous
+- `category`: Already exiting in `Libobject`, used for import, could be also used by `Search`.
+
+A draft implementation or Part I is available in PR
+https://github.com/coq/coq/pull/16484
+
+This changes affect the lifetime in the following way:
+
+## Creation
+
+Clients must provide the metadata when registering the objects.
+
+## Loading
+
+`Declaremods` already stores the objects in a map indexed by their
+prefix, we propose to:
+
+- create a reverse map or maps to allow clients to query the `Data.t`
+  efficiently. In particular we could have reverse maps for
+  categories, names, locations. For now we implement a reverse map for
+  names, but more options could be desirable after more experimentation.
+
+- expose the regular map for anonymous objects so clients can explore.
+
+## Extra
+
+It would be desirable to allow clients to register as to allow
+incremental indexing of objects, of for example, to maintain a
+location-based table.
+
+Another interesting topic is versioning of objects. While preliminary,
+we could include versioning information in the objects as for example
+to allow to have hybrid Coq 8/9 objects, or several versions of the
+same `Definition` for example.
+
+# Part II: behavior specification
+
+The second part of this CEP, not yet implemented, is to phase out the
+`declare_object` API in favor a set of `Behavior.t` traits that an
```

Comment:
Is there a description of the `Behavior` API?

Reply:
I don't have a too precise view on how to implement it, there are several options so here we can use some feedback.

I guess for the first prototype I'd try to have a list of traits, indeed due to the existential nature of traits that would mean turning the declaration of objects into a module, but I think I'm fine with that (as in the case for errors)